pkg/controller/util/endpoint/rate_limiting_queue.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package endpoint
+
+import (
+	"sync"
+
+	"k8s.io/client-go/util/workqueue"
+)
+
+// RateLimitingQueue implements a wrapper on workqueue.RateLimitingInterface that also rate limits initial key additions
+// to the queue (unlike workqueue.RateLimitingInterface which rate limits only error retries).
+// TODO(mborsz): Rename to something that doesn't collide with workqueue.RateLimitingInterface (or refactor workqueue.RateLimitingInterface).
+// TODO(mborsz): Promote to workqueue package once we validate API in endpoint controllers.
+type RateLimitingQueue struct {
```

Comment:
Can you add a guard that it implements RateLimiitingInterface

var _ RateLimitingInterface = &RateLimitingQueue{}

Reply:
But it doesn't implement RateLimitingInterface: It doesn't implement Add, AddRateLimited methods.

The reason for this is that AddRateLimited method which is required by RateLimitingInterface is quite confusing in this context:
- In RateLimitingInterface, the meaning of AddRateLimiting is that we are retrying some element and error rate limiter should be used (there is no rate limiter for initial additions to the queue)
- In our struct, we have two "rate limitings": initial rate limiting and error rate limiting (implemented by workqueue.RateLimitingInterface) so for potential user of this struct it wouldn't be clear what AddRateLimited means in this context.